File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
# class Solution:
#     def subarraySum(self, nums: List[int], k: int) -> int:
#         n = len(nums)
#         count , sum_ = 0 , 0
#         my_map = defaultdict(int)
        
#         for num in nums :
#             sum_ += num
#             if sum_ == k :
#                 count += 1
#             if (sum_ - k) in my_map :
#                 count += my_map[sum_ - k]
                
#             my_map[sum_]+=1
            
#         return count
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        count , sum_ = 0 , 0
        counter = Counter()

        for num in nums :
            sum_ += num
            if sum_ == k :
                count += 1
            if counter[sum_ - k] != 0 :
                count += counter[sum_ - k]
            counter[sum_] += 1
                
        return count

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int count = 0 , sum = 0;
        HashMap<Integer,Integer> mpp = new HashMap<>();
        
        for(int i=0;i<nums.length;i++){
            sum += nums[i];
            if(sum == k) count += 1;
            if(mpp.get(sum - k) != null) count += mpp.get(sum - k);
            mpp.put(sum,mpp.getOrDefault(sum,0)+1);
        }
        return count;
    }
}